walker.go

```diff
@@ -0,0 +1,429 @@
+
+// errPauseWalkOperation signals the pause of the walk operation.
+var errPauseWalkOperation = errors.New("pause in the current walk operation")
+
+// ErrNilVisitor signals the lack of a `Visitor` function.
+var ErrNilVisitor = errors.New("no Visitor function specified")
+
+// Iterate the DAG through the DFS pre-order walk algorithm, going down
+// as much as possible, then `Next` to the other siblings, and then up
+// (to go down again). The position is saved throughout iterations (and
+// can be previously set in `Search`) allowing `Iterate` to be called
+// repeatedly (after a `Pause`) to continue the iteration.
+//
+// This function returns the errors received from `down` (generated either
+// inside the `Visitor` call or any other errors while fetching the child
+// nodes), the rest of the move errors are handled within the function and
+// are not returned.
+func (w *Walker) Iterate(visitor Visitor) error {
+
+	// Iterate until either: the end of the DAG (`errUpOnRoot`), a `Pause`
+	// is requested (`errPauseWalkOperation`) or an error happens (while
+	// going down).
+	for {
+
+		// First, go down as much as possible.
+		for {
+			err := w.down(visitor)
+
+			if err == ErrDownNoChild {
+				break
+				// Can't keep going down from this node, try to move Next.
+			}
+
+			if err == errPauseWalkOperation {
+				return nil
+				// Pause requested, `errPauseWalkOperation` is just an internal
+				// error to signal to pause, don't pass it along.
+			}
+
+			if err != nil {
+				return err
+				// `down` is the only movement that can return *any* error.
+			}
+		}
+
+		// Can't move down anymore, turn to the next child in the `ActiveNode`
+		// to go down a different path. If there are no more child nodes
+		// available, go back up.
+		for {
+			err := w.Next()
+			if err == nil {
+				break
+				// No error, it turned to the next child. Try to go down again.
+			}
+
+			// It can't go Next (`ErrNextNoChild`), try to move up.
+			err = w.up()
+			if err != nil {
+				// Can't move up, on the root again (`errUpOnRoot`).
+				return EndOfDag
+			}
+
+			// Moved up, try `Next` again.
+		}
+
+		// Turned to the next child (after potentially many up moves),
+		// try going down again.
+	}
+}
+
+// Search a specific node in a downwards manner. The `Visitor` should be
+// used to steer the search selecting at each node which child will the
+// search continue to (extending the `path` in that direction) or pause it
+// (if the desired node has been found). The search always starts from
+// the root. It modifies the position so it shouldn't be used in-between
+// `Iterate` calls (it can be used to set the position *before* iterating).
+// If the visitor returns any non-`nil` errors the search will stop.
+//
+// TODO: The search could be extended to search from the current position.
+// (Is there something in the logic that would prevent it at the moment?)
+func (w *Walker) Search(visitor Visitor) error {
```

Comment:
I think I’d say the same thing as I did before about this name — `Seek()` sounds a lot clearer to me because what you’re really doing here is navigating a direct path to a leaf. It only works as a search if the the field you are searching on is indexed in every node of the tree. For example, you can search for a leaf by byte offset in a UnixFS tree with this, but if I want to search for the text “hello” in a text file in a UnixFS tree, this function won’t be useful.

Second, since this still calls the visitor on the leaf node, I feel like I’m back in the position where it doesn’t feel like this offers much over just calling `Iterate()` like I described here: https://github.com/ipfs/go-ipfs/pull/5257#issuecomment-414714047 The only difference between what you’d write with `Iterate()` vs. `Search()` is that you’d have to call `Pause()` before returning on a leaf (and you’d already be having to check whether you’re on a leaf anyway). The logic in your visitor would be the same in both cases. Why not make this function do a little more?

```go
func (w *Walker) Seek(visitor Visitor) (NavigableNode, error) {
    err := w.Iterate(func(node NavigableNode) error {
        // If we’ve arrived at a leaf, we’re done! \o/
        if node.ChildTotal() == 0 {
            return errPauseWalkOperation
        }
        err := visitor(node)
        if err == ErrNextNoChild || w.ActiveChildIndex() >= w.ActiveNode().ChildTotal() {
            // You probably want a special, public error for this case, but this
            // reproduces what the current commit does.
            return errPauseWalkOperation
        }
        return err
    })
    return w.ActiveNode(), err
}
```

That way, the visitor you use with `Search()` (or `Seek()` or whatever) can have simpler logic — it only has to navigate and not worry about processing the final node.

Reply:
> For example, you can search for a leaf by byte offset in a UnixFS tree with this, but if I want to search for the text “hello” in a text file in a UnixFS tree, this function won’t be useful.

I'll rename it to `Seek` if you think it's more appropriate, but just to clarify, this doesn't operate on the UnixFS layer (or any particular IPFS layer for that matter), it operates on the more abstract concept of a DAG, *any* DAG (not just the MerkleDAG). What you search/seek for depends on what those generic (not necessarily IPLD) nodes encode, it might be an offset but it might as well be a string, it's the consumer of the walker who defines what information is encoded and how it's processed.

Reply:
> Second, since this still calls the visitor on the leaf node, I feel like I’m back in the position where it doesn’t feel like this offers much over just calling `Iterate()` like I described here: [ipfs/go-ipfs#5257 (comment)](https://github.com/ipfs/go-ipfs/pull/5257#issuecomment-414714047)

Yes, and I'm still giving more importance to what you (very well) described:

> But you’re right; if we changed the API to Search, we could do something like this, that might be conceptually clearer (at the cost of more error handling code):

But you have a good point, I'm just prioritizing the concept over the code that relies inside the walker (which the user should not be forced to study to use it).




> The only difference between what you’d write with `Iterate()` vs. `Search()` is that you’d have to call `Pause()` before returning on a leaf (and you’d already be having to check whether you’re on a leaf anyway).

I'm not sure I understand this part, the advantage with `Search()` is that the check for a leaf node (i.e., nowhere to go from here) is done inside the walker method so the user doesn't need to worry about it. In contrast, using `Iterate()` to perform a search/seek lays that burden on the user (granted, small burden, but still if I can avoid it without making too much of a mess inside the walker I think it's worth it).

I think that what may be confusing is the way the DAG reader uses the walker when [seeking inside a file](https://github.com/ipfs/go-unixfs/pull/12/files#diff-0688519bad7de835291edaa3bf6d6e34R305). It does check for leaf nodes, but not because of  the DAG seek operation itself but because of the way UnixFS works, which encodes information in the leaf nodes differently than in internal nodes. But those two are decoupled concepts.

Reply:
> just to clarify, this doesn't operate on the UnixFS layer (or any particular IPFS layer for that matter)

Yep, that was clear. I guess I could have said “search for some arbitrary string of bytes” instead 😉

> I'm not sure I understand this part, the advantage with `Search()` is that the check for a leaf node (i.e., nowhere to go from here) is done inside the walker method

Maybe I’m misreading, then. The code looks to me like the user-supplied `Visitor` function will be called with the leaf. It [just keeps going down until there’s no child to enter](https://github.com/ipfs/go-ipld-format/pull/39/files#diff-0134f9924b5fee98958ef242d42de007R255). It seems like it’s still up to the user to determine whether their visitor should be path-seeking or consuming the found node on any given call.

Reply:
> The code looks to me like the user-supplied `Visitor` function will be called with the leaf. It [just keeps going down until there’s no child to enter](https://github.com/ipfs/go-ipld-format/pull/39/files#diff-0134f9924b5fee98958ef242d42de007R255).

Yes.



> It seems like it’s still up to the user to determine whether their visitor should be path-seeking or consuming the found node on any given call.

Could you expand on that? What does it mean to be path-seeking? The way I thought of it (but this may not be the actual use case) is that you're presented with the choice on every bifurcation, either you found your node and stop or you use the current node to steer the seek/search down. There's no need to worry about the special leaf node case, the search would just stop on its own and you'd end up with nothing (no node found). If using `Iterate` the user must supply the "is this the leaf node" case, and then issue another pause to stop the search (but not because they found what they we're looking for but because there's nowhere to go).

Reply:
> What does it mean to be path-seeking?

Running logic that determines where to go in the DAG (i.e. determining whether and how many times to call `walker.NextChild()`).

> The way I thought of it (but this may not be the actual use case) is that you're presented with the choice on every bifurcation, either you found your node and stop or you use the current node to steer the seek/search down.

I guess what I’m stuck on is two things:

1. If you are on a leaf, you are by definition already done. Why should the visitor have to deal with that leaf?

2. We know it’s common to have leaves that are differently formatted than intermediate nodes (UnixFS is a great example). If that’s true, then your visitor has to be more complex:

    ```go
    err = walker.Seek(func(node NavigableNode) error {
        // Figure out how to treat this node
        if node.ChildTotal() > 0 {
            // Do some seeking: determine whether to skip children, etc.
            // call `walker.NextChild()` as needed
        } else {
            // Handle the found leaf node node
            // Even if I don't care about this case and don't need the `else`
            // here, I still need the `if` above
        }
        return nil
    })
    ```

    But an implementation that never sends the leaf to the Visitor lets the Visitor be much simpler and only worry about one thing:

    ```go
    err = walker.Seek(func(node NavigableNode) error {
        // Do some seeking: determine whether to skip children, etc.
        // call `walker.NextChild()` as needed
        return nil
    })
    ```

> If using `Iterate` the user must supply the "is this the leaf node" case, and then issue another pause to stop the search

So my concern above is that, with this `Seek` implementation, they still have to to supply the “is this the leaf node” case. It *feels* to me like, if the only difference is that I don’t have to explicitly call `Pause()`, I haven’t gained much. To bring back a [slightly modified] comparison from the original comment on the other PR:

    ```go
    err = walker.Seek(func(node NavigableNode) error {
        if node.ChildTotal > 0 {
            // Do some seeking: determine whether to skip children, etc.
            // call `walker.NextChild()` as needed
        }
        return nil
    })

    # Achieving the same functionality with Iterate:
    err = walker.Iterate(func(node NavigableNode) error {
        if node.ChildTotal() > 0 {
            // Do some seeking: determine whether to skip children, etc.
            // call `walker.NextChild()` as needed
        } else {
            walker.Pause()
        }
        return nil
    })
    ```

Reply:
> So my concern above is that, with this `Seek` implementation, they still have to to supply the “is this the leaf node” case. It _feels_ to me like, if the only difference is that I don’t have to explicitly call `Pause()`, I haven’t gained much.


(The next explanation is not that important, I'm fine going either way with this API).

Those are two different uses of the check `node.ChildTotal() > 0`. UnixFS (and probably many more) use a special node type for the leaf and the implementation will need to check for that, but (ideally) that should be handled in the UnixFS implementation and not here (e.g., UnixFS should provide a function to tell if the current node contains the target offset, is lower or higher).

For example, I had to introduce a [function](https://github.com/ipfs/go-unixfs/pull/12/files#diff-bcd050cb91d8c2b8d1366221e53ce797R311) in UnixFS to extract file data independently of the node type so it becomes more transparent to the consumer and that way the `Visitor` passed to  `Iterate` in the [DAG read method](https://github.com/ipfs/go-unixfs/pull/12/files#diff-0688519bad7de835291edaa3bf6d6e34R170) doesn't need to care about the leaf node check (I added it anyway as a performance improvement but it might as well not be there).

I'm not sure if we'll introduce a similar function to seek offsets and avoid the `node.ChildTotal() > 0` check but decoupling those uses helps clarify (I hope) who is doing what and why. Sometimes for performance reasons we collapse those use cases but that normally makes the code harder to understand.